### PR TITLE
ANCM Test Refactoring: Initial check-in to fix the build issue on AspNetCI

### PR DESCRIPTION
**Pull request by jhkimnew**
@pan-wang @jkotalik 
This is the initial test code fix for ASPNet CI. 
Experimentally, I set only two test cases with P0 so that I can verify those tests got passed in CI build machine. I will set P0 for other test cases later.

### CODE REVIEW

**Comment by jkotalik** [comment]
@jhkimnew please sign https://cla2.dotnetfoundation.org/ so the cla-required label goes away 😄

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 1]
```diff
@@ -5,34 +5,52 @@
...
+    {
+        public const string TestSkipContext = "SkipTest";
+        public const string UsePrivateANCM = "UsePrivateANCM";
+        public const string UseIISExpressContext = "UseIISExpress";
+        public const string UseFullIISContext = "UseFullIIS";
+        public const string RunAsNonAdminUser = "RunAsNonAdminUser";
+        public const string MakeCertExeAvailable = "MakeCertExeAvailable";
+        public const string X86Platform = "X86Platform";
+        public const string Wow64BitMode = "Wow64BitMode";
+        public const string P0 = "P0";
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: P0 and P1 are odd names for privilege levels, @shirhatti can we change this?

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 2]
```diff
@@ -5,34 +5,52 @@
...
+
+        public const string ANCMTestFlags_DefaultContext_IISExpressP0Only = ANCMTestFlags.UsePrivateANCM
+                                                        + ";"
+                                                        + ANCMTestFlags.UseIISExpressContext
+                                                        + ";"
+                                                        + ANCMTestFlags.RunAsNonAdminUser
+                                                        + ";"
+                                                        + ANCMTestFlags.P0;
+
+        public const string ANCMTestFlags_DefaultContext_FullIIS = ANCMTestFlags.UsePrivateANCM
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: Does the full IIS test require using the private ANCM?

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 3]
```diff
@@ -5,34 +5,52 @@
...
+                                                        + ";"
+                                                        + ANCMTestFlags.P0
+                                                        + ";"
+                                                        + ANCMTestFlags.P1;
 
         private static bool? _usePrivateAspNetCoreFile = null;
         public static bool? UsePrivateAspNetCoreFile
         {
             get {
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: nit: move open brace to next line.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 4]
```diff
@@ -65,27 +83,135 @@ public class InitializeTestMachine : IDisposable
         private static bool _InitializeTestMachineCompleted = false;
         private string _setupScriptPath = null;
         
-        private bool CheckPerquisiteForANCMTest()
+        private static bool? _MakeCertExeAvailable = null;
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: nit: lowercase m.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 5]
```diff
@@ -65,27 +83,135 @@ public class InitializeTestMachine : IDisposable
...
+
+        private static string _globalTestFlags = null;
+        public static string GlobalTestFlags
+        {
+            get
+            {
+                if (_globalTestFlags == null)
+                {
+                    bool isElevated;
+                    WindowsIdentity identity = WindowsIdentity.GetCurrent();
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: These lines can be condense significantly. Ex identity is only used on the next line, so just pass in WindowsIdentity.GetCurrent() into it.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 6]
```diff
@@ -65,27 +83,135 @@ public class InitializeTestMachine : IDisposable
...
+                    WindowsPrincipal principal = new WindowsPrincipal(identity);
+                    isElevated = principal.IsInRole(WindowsBuiltInRole.Administrator);
+
+                    // check if this test process is started with the Run As Administrator start option
+                    _globalTestFlags = Environment.ExpandEnvironmentVariables(ANCMTestFlagsEnvironmentVariable);
+                    _globalTestFlags = _globalTestFlags.ToLower();
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: This may be for a different PR, but instead of lower casing all of the strings, we can maybe compare some of them ignorecase?

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 7]
```diff
@@ -65,27 +83,135 @@ public class InitializeTestMachine : IDisposable
...
+                    if (_globalTestFlags.Contains(ANCMTestFlags.UseIISExpressContext.ToLower()) && _globalTestFlags.Contains(ANCMTestFlags.UseFullIISContext.ToLower()))
+                    {
+                        throw new ApplicationException(ANCMTestFlagsEnvironmentVariable + " can't be set with both "
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: Probably not a good idea to be throwing exceptions from properties.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 8]
```diff
@@ -65,27 +83,135 @@ public class InitializeTestMachine : IDisposable
...
+                    {
+                        _globalTestFlags = ANCMTestFlags_DefaultContext_FullIIS;
+                    }
+
+                    // adjust the default test context in run time to figure out wrong test context values
+                    if (!isElevated)
+                    {
+                        if (_globalTestFlags.Contains(ANCMTestFlags.UseFullIISContext.ToLower()))
+                        {
+                            _globalTestFlags = _globalTestFlags.Replace(ANCMTestFlags.UseFullIISContext.ToLower(), "");
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: I'm not a fan of calling Replace like this. Maybe go back and build up the string based on flags?

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 9]
```diff
@@ -65,27 +83,135 @@ public class InitializeTestMachine : IDisposable
...
+                        }
+
+                        if (!_globalTestFlags.Contains(ANCMTestFlags.RunAsNonAdminUser.ToLower()))
+                        {
+                            TestUtility.LogInformation("Added test context of " + ANCMTestFlags.RunAsNonAdminUser);
+                            _globalTestFlags += ";" + ANCMTestFlags.RunAsNonAdminUser;
+                        }
+                    }
+
+                    if (MakeCertExeAvailable == true)
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: nit: remove true check.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 7 above)
Comment (jkotalik) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: Maybe make this a method?

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 10]
```diff
@@ -65,27 +83,135 @@ public class InitializeTestMachine : IDisposable
...
+                    _globalTestFlags = _globalTestFlags.ToLower();
+                }
+
+                return _globalTestFlags;
+            }            
+        }
+
         public InitializeTestMachine()
         {
             _referenceCount++;
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: Can you break up this method into multiple methods? I'm having trouble understanding what it is doing.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 11]
```diff
@@ -177,10 +296,14 @@ public InitializeTestMachine()
                     TestUtility.LogInformation("We will use IISExpress instead of IIS: " + ex.Message);
                 }
 
-                string siteRootPath = Path.Combine(Environment.ExpandEnvironmentVariables("%SystemDrive%") + @"\", "inetpub", "ANCMTest");
+                string siteRootPath = TestRootDirectory;
                 if (!Directory.Exists(siteRootPath))
                 {
                     Directory.CreateDirectory(siteRootPath);
+                    DirectorySecurity sec = Directory.GetAccessControl(siteRootPath);
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: Will this be hit by running IIS express? If so, it requires an authenticated user, which I think needs Admin permissions?

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 12]
```diff
@@ -210,7 +333,7 @@ public InitializeTestMachine()
                 
                 if (InitializeTestMachine.UsePrivateAspNetCoreFile == true)
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: nit: remove == true.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 13]
```diff
@@ -289,7 +412,7 @@ private void RollbackIISApplicationhostConfigFile()
             }
         }
 
-        private void PreparePrivateANCMFiles()
+        private void PreparePrivateANCMFiles(bool isIISReady)
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: This bool is only used once, just read IISConfigUtility.IsIISReady.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 14]
```diff
@@ -323,30 +446,40 @@ private void PreparePrivateANCMFiles()
...
-                        TestUtility.ResetHelper(ResetHelperMode.StopW3svcStartW3svc);
-                        Thread.Sleep(1000);
+                        //
+                        // NOTE: ANCM schema file can't be overwritten here, if there is any schema change, that should be updated with installing setup
+                        //
 
-                        string from = Path.Combine(outputPath, "x64", "aspnetcore.dll");
-                        TestUtility.FileCopy(from, FullIisAspnetcore_path, overWrite:true, ignoreExceptionWhileDeletingExistingFile:false);
-                        TestUtility.FileCopy(from, IisExpressAspnetcore_path, overWrite: true, ignoreExceptionWhileDeletingExistingFile: false);
+                        if (isIISReady)
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: I'm confused by this check. If IIS is ready, we wait?

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 15]
```diff
@@ -323,30 +446,40 @@ private void PreparePrivateANCMFiles()
...
-                            from = Path.Combine(outputPath, "Win32", "aspnetcore.dll");
-                            TestUtility.FileCopy(from, FullIisAspnetcore_X86_path, overWrite: true, ignoreExceptionWhileDeletingExistingFile: false);
-                            TestUtility.FileCopy(from, IisExpressAspnetcore_X86_path, overWrite: true, ignoreExceptionWhileDeletingExistingFile: false);
+                            string from_32bit = Path.Combine(outputPath, "Win32", "aspnetcore.dll");
 
-                            // NOTE: schema file can't be overwritten, if there is any schema change, that should be updated manually
-                            from = Path.Combine(outputPath, "Win32", "aspnetcore_schema.xml");
-                            TestUtility.FileCopy(from, IisExpressAspnetcoreSchema_X86_path, overWrite: false, ignoreExceptionWhileDeletingExistingFile: false);
-                        }
+                            if (isIISReady)
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: There are a lot of checks for that bool. Consider separating into two methods rather than checking it multiple times.

File: test/AspNetCoreModule.Test/Framework/TestUtility.cs [hunk 16]
```diff
@@ -263,6 +263,7 @@ public static void LogInformation(string format, params object[] parameters)
         {
             if (format != null)
             {
+                //Logger.LogInformation("[" + DateTime.Now.ToString() + "] " + format, parameters);
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/TestUtility.cs]: nit remove this.

File: test/AspNetCoreModule.Test/Framework/TestWebSite.cs [hunk 17]
```diff
@@ -308,10 +327,30 @@ public TestWebSite(IISConfigUtility.AppPoolBitness appPoolBitness, string logger
...
+                            iisConfig.AddModule("AspNetCoreModule", (InitializeTestMachine.IisExpressAspnetcore_path), null);
+                        }
+                    }
+                    else
+                    {
+                        iisConfig.AddModule("AspNetCoreModule", (InitializeTestMachine.IisExpressAspnetcore_path), null);
+                    }
+                }
+
+                string hostname = "";
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/TestWebSite.cs]: This is repeated twice, abstract to a new method.

File: test/AspNetCoreModule.Test/FunctionalTestHelper.cs [hunk 18]
```diff
@@ -23,64 +23,48 @@
 namespace AspNetCoreModule.Test
 {
     [AttributeUsage(AttributeTargets.Method, AllowMultiple = true)]
-    public class ANCMTestSkipCondition : Attribute, ITestCondition
+    public class ANCMTestTestPriority : Attribute, ITestCondition
     {
-        private readonly string _environmentVariableName;
-        public ANCMTestSkipCondition(string environmentVariableName)
+        private readonly string _value;
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/FunctionalTestHelper.cs]: nit: what is value here?

File: test/AspNetCoreModule.Test/FunctionalTestHelper.cs [hunk 19]
```diff
@@ -23,64 +23,48 @@
...
-                        if (token.Equals(InitializeTestMachine.ANCMTestFlagsTestSkipContext, StringComparison.InvariantCultureIgnoreCase))
+                        if (token.Equals(ANCMTestFlags.TestSkipContext, StringComparison.InvariantCultureIgnoreCase))
                         {
-                            AdditionalInfo = InitializeTestMachine.ANCMTestFlagsTestSkipContext + " is set";
+                            testSkipFlagFound = true;
                             result = false;
+                            break;
+                        }
+                    }
+                    if (result == false)
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/FunctionalTestHelper.cs]: nit !result

File: test/AspNetCoreModule.Test/FunctionalTestHelper.cs [hunk 20]
```diff
@@ -136,11 +122,12 @@ public enum ReturnValueType
                 {
                     BaseAddress = testSite.AspNetCoreApp.GetUri(),
                     Timeout = TimeSpan.FromSeconds(5),
-                };
+                };
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/FunctionalTestHelper.cs]: nit remove extra space.

**Review by jkotalik — CHANGES_REQUESTED** [review]
- We prefer to use var instead of type name for variable declaration.
- I ran these tests locally and only 6 tests ran. We are changing the functionality such that running in an Administrator terminal doesn't run all tests?

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 1 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: Okay, updated 
P0 -> Default
P1 -> RequireRunAsAdministrator

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 2 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: There is no restriction. Users can choose whether the private file is used or not for full IIS test. However, if the test is started on non-admin user, test will use IISExpress automatically and in that case there is no other choice.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 3 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: fixed

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 4 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: fixed

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 5 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: fixed

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 6 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: I will revisit this issue when other work items are finished.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 7 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: This is fixed.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 8 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: I agree but I will handle this in the future.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 9 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: Fixed

Comment (jkotalik) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: No like just if(MakeCertExeAvailable)

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 10 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: Fixed

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 12 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: fixed

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 13 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: fixed

File: test/AspNetCoreModule.Test/FunctionalTestHelper.cs (hunk 20 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/FunctionalTestHelper.cs]: Fixed

File: test/AspNetCoreModule.Test/FunctionalTestHelper.cs (hunk 19 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/FunctionalTestHelper.cs]: Fixed

File: test/AspNetCoreModule.Test/FunctionalTestHelper.cs (hunk 18 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/FunctionalTestHelper.cs]: Fixed with a specific name.

File: test/AspNetCoreModule.Test/Framework/TestUtility.cs (hunk 16 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/TestUtility.cs]: Fixed

File: test/AspNetCoreModule.Test/Framework/TestWebSite.cs (hunk 17 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/TestWebSite.cs]: Ah, good catch!!! Fixed.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 14 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: This is not for waiting. Just double-check if the private file is successfully copied and this is only for Full IIS Server. Code is now updated to make it more clear.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 15 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: Fixed.

**Comment by jhkimnew** [comment]
@jkotalik Update is done. Please approve so that I can complete this PR.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 9 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: This is already fixed. I think you looked at the old version before I push the new change.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 11 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs]: This does not require Admin permission. I already tested it.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 21]
```diff
@@ -65,122 +48,260 @@ public class InitializeTestMachine : IDisposable
...
+                    try
+                    {
+                        string makecertExeFilePath = TestUtility.GetMakeCertPath();
+                        TestUtility.RunCommand(makecertExeFilePath, null, true, true);
+                        TestUtility.LogInformation("Verified makecert.exe is available : " + makecertExeFilePath);
+                        _makeCertExeAvailable = true;
+                    }
+                    catch
+                    {
+                        _makeCertExeAvailable = false;
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:68]: You don't need to set _makeCertExeAvailable to false here.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 22]
```diff
@@ -65,122 +48,260 @@ public class InitializeTestMachine : IDisposable
...
+                    }
+                }
+                return (_makeCertExeAvailable == true);
             }
-            return result;
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:79]: Just return _makeCertExeAvailable

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 23]
```diff
@@ -65,122 +48,260 @@ public class InitializeTestMachine : IDisposable
...
+        {
+            get
+            {
+                if (_globalTestFlags == null)
                 {
-                    System.Diagnostics.Debugger.Launch();                    
-                }
+                    bool isElevated;
+                    WindowsPrincipal principal = new WindowsPrincipal(WindowsIdentity.GetCurrent());
+                    isElevated = principal.IsInRole(WindowsBuiltInRole.Administrator);
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:92]: No need to declare isElevated before setting it here.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 24]
```diff
@@ -65,122 +48,260 @@ public class InitializeTestMachine : IDisposable
...
+                isIISInstalled = false;
+            }
+            
+            if (!isIISInstalled)
+            {
+                throw new System.ApplicationException("IIS server is not installed");
+            }
+
+            // Check websocket is installed
+            if (File.Exists(Path.Combine(IISConfigUtility.Strings.IIS64BitPath, "iiswsock.dll")))
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:220]: The win7 check for websockets occur on each test? This would fail if we were running tests on win7.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 25]
```diff
@@ -65,122 +48,260 @@ public class InitializeTestMachine : IDisposable
...
+            {
+                TestUtility.LogInformation("Websocket is installed");
+            }
+            else
+            {
+                throw new System.ApplicationException("websocket module is not installed");
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:226]: Add "using System" to the top of the file and say "throw new ApplicationException("...");" for all of these.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 26]
```diff
@@ -65,122 +48,260 @@ public class InitializeTestMachine : IDisposable
...
+            {
+                TestUtility.LogInformation("W3SVC service is restarted after restoring applicationhost.config");
+            }
+            else
+            {
+                throw new System.ApplicationException("WWW service can't start");
+            }
+
+            // check URLRewrite module exists
+            if (File.Exists(Path.Combine(IISConfigUtility.Strings.IIS64BitPath, "rewrite.dll")))
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:248]: For my own knowledge, what does the UrlRewrite module need to be installed for ANCM?

Comment (jkotalik) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:248]: why*

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs [hunk 27]
```diff
@@ -65,122 +48,260 @@ public class InitializeTestMachine : IDisposable
...
+
+            if (IISConfigUtility.ApppHostTemporaryBackupFileExtention == null)
+            {
+                throw new System.ApplicationException("Failed to backup applicationhost.config");
+            }
+        }
+
+        public InitializeTestMachine()
+        {
+            _referenceCount++;
```
Comment (jkotalik) [inline test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:265]: This is a weird way of doing the singleton pattern. Maybe just have a bool instead or make a new object?

**Review by jkotalik — COMMENTED** [review]
I'll look more into functional components later.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 21 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:68]: Fixed and will send a new commit soon.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 22 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:79]: This is by design because the type of MakeCertExeAvalialbe is bool, which is different to _makeCertExeAvailable's (bool?).

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 23 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:92]: Fixed and will send a new commit soon.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 24 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:220]: Fixed and will send a new commit soon. I made a new test flag of WebSocketModuleAvailable. I will make the websocket test cases requires the flag.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 25 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:226]: Fixed and will send a new commit soon.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 26 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:248]: Fixed and will send a new commit soon. I made a new test flag of UrlRewriteModuleAvailable. I will make the related test cases to use the flag.
When the urlrewrite rule is enabled, the request sent to IIS is create a new child request based on the UrlRewrite rule. And the child request should be tested to check if it works with ANCM as well.

File: test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs (hunk 27 above)
Comment (jhkimnew) [reply test/AspNetCoreModule.Test/Framework/InitializeTestMachine.cs:265]: I will revisit this and remove this. Eventually this will be gone because the singleton logic won't be necessary when the code refactoring is done.

**Comment by jhkimnew** [comment]
@jkotalik Hey, I sent a new commit. Please approve so that I can complete this and check this new change can fix the test run skip issue on AspnetCI agent machine.

**Review by jkotalik — APPROVED** [review]

**Comment by jkotalik** [comment]
@jhkimnew I'm working on getting ANCM to run CI tests when you make a pull request rather than just when you push to dev. That way we don't need to verify when merging and rather can check before.

**Comment by jhkimnew** [comment]
+ @pan-wang, @jkotalik  FYI Justin, the build-check on CI when PR is created, not when PR is completed, will be a nice feature for us. Thanks for working on that. 👍

**Comment by shirhatti** [comment]
From a cursory glance, looks like AppVeyor agents have IIS Express installed.
@jkotalik Can we look at enabling these tests on AppVeyor as well?
